x/ccv/provider/keeper/partial_set_security.go

```diff
@@ -0,0 +1,36 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/interchain-security/v4/x/ccv/provider/types"
+)
+
+type OptedInValidator struct {
+	ProviderAddr types.ProviderConsAddress
+	// block height the validator opted in at
+	BlockHeight uint64
+}
+
+func (k Keeper) HandleOptIn(ctx sdk.Context, chainID string, providerAddr types.ProviderConsAddress, consumerKey *string) {
+	if k.IsToBeOptedOut(ctx, chainID, providerAddr) {
+		// a validator to be opted in cancels out with a validator to be opted out
+		k.DeleteToBeOptedOut(ctx, chainID, providerAddr)
+	} else if !k.IsToBeOptedIn(ctx, chainID, providerAddr) && !k.IsOptedIn(ctx, chainID, providerAddr) {
+		// a validator can only be set for opt in if it is not opted in and not already set for opt in
+		k.SetToBeOptedIn(ctx, chainID, providerAddr)
+	}
+
+	if consumerKey != nil {
+		// TODO: assign consumer key in this case
+	}
+}
+
+func (k Keeper) HandleOptOut(ctx sdk.Context, chainID string, providerAddr types.ProviderConsAddress) {
```

Comment:
Same thing here, it should return an error if the consumer doesn't exist as explained above.